[UI-side compositing] Keyboard scrolling Amazon.com crashes Safari

rr-codes opened the pull request:
#### 40453f0edcf34cb76ca4095f3a779a69a09c8393
<pre>
[UI-side compositing] Keyboard scrolling Amazon.com crashes Safari
<a href="https://bugs.webkit.org/show_bug.cgi?id=248924">https://bugs.webkit.org/show_bug.cgi?id=248924</a>
rdar://103058005

Reviewed by Aditya Keerthi.

`KeyboardScroll` and `RequestedKeyboardScrollData` were not IPC encodable or
decodable, so they were initialized with arbitrary values with UI-side compositing,
causing unexpected behavior and crashing.

This PR makes both of these encodable and decodable, provides equality comparison
operators for both, and provides all fields of both data types with sensible
initial values.

* Source/WebCore/page/scrolling/ScrollingCoordinatorTypes.h:
(WebCore::RequestedKeyboardScrollData::operator== const):
* Source/WebCore/platform/KeyboardScroll.h:
(WebCore::KeyboardScroll::operator== const):
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.cpp:
(ArgumentCoder&lt;KeyboardScroll&gt;::encode):
(ArgumentCoder&lt;KeyboardScroll&gt;::decode):
(WebKit::dump):

Canonical link: <a href="https://commits.webkit.org/257734@main">https://commits.webkit.org/257734@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ead2d33e26b6a6b87269be72b74f7b3aa7fb446b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99530 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8706 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32625 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108896 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169131 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/103523 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/9291 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/86013 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92006 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106813 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105285 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/7019 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90548 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33990 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/88844 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21900 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/77261 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2552 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23414 "Passed tests") | | 
| [⏳ 🛠 🧪 jsc-arm64 ](https://ews-build.webkit.org/#/builders/JSC-Tests-arm64-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2482 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/45803 "Build is being retried. Recent messages:validate-change running; Cleaned up git repository; 'git checkout ...'; Skipping applying patch since patch_id isn't provided; Checked out pull request; Pull request contains relevant changes; Skipped layout-tests; Ignored 1 pre-existing failure based on results-db; triggering crash log submission") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/8634 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42889 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5315 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4341 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.cpp:
```diff
@@ -557,6 +568,46 @@ bool ArgumentCoder<SnapOffset<float>>::decode(Decoder& decoder, SnapOffset<float
 }
 
 
+void ArgumentCoder<KeyboardScroll>::encode(Encoder& encoder, const KeyboardScroll& scroll)
```
We should be able to use `WebCoreArgumentCoders.serialization.in` rather than manually writing the encoder/decoder.

rr-codes replied:
@pxlcoder ah ok, didn't know that was a thing! I update my PR, lmk if I did it correctly!

pxlcoder replied:
It's a relatively new addition 😃

smfr commented on Source/WebCore/page/scrolling/ScrollingCoordinatorTypes.h:
```diff
@@ -140,8 +140,13 @@ enum class KeyboardScrollAction : uint8_t {
-    KeyboardScrollAction action;
-    std::optional<KeyboardScroll> keyboardScroll;
+    KeyboardScrollAction action { KeyboardScrollAction::StartAnimation };
+    std::optional<KeyboardScroll> keyboardScroll { };
```
No need to explicitly initialize a `std::optional<>`. It has its own constructor.

smfr commented on Source/WebCore/platform/KeyboardScroll.h:
```diff
@@ -33,12 +33,21 @@ namespace WebCore {
-    FloatSize offset; // Points per increment.
-    FloatSize maximumVelocity; // Points per second.
-    FloatSize force;
+    FloatSize offset { }; // Points per increment.
```
No need to initialize classes that have constructors.

pxlcoder commented on Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
```diff
@@ -1768,12 +1768,26 @@ header: <WebCore/ScrollingCoordinatorTypes.h>
     bool useDarkAppearanceForScrollbars;
 };
 
+[CustomHeader] struct WebCore::RequestedKeyboardScrollData {
```
Nit: Let's add `header: <WebCore/KeyboardScroll.h>` above this line for consistency.

rr-codes replied:
@pxlcoder but `RequestedKeyboardScrollData` isn't in `KeyboardScroll.h` ?

pxlcoder replied:
sorry, `header: <WebCore/ScrollingCoordinatorTypes.h>`

rr-codes committed:
[UI-side compositing] Keyboard scrolling Amazon.com crashes Safari
https://bugs.webkit.org/show_bug.cgi?id=248924
rdar://103058005

Reviewed by Aditya Keerthi.

`KeyboardScroll` and `RequestedKeyboardScrollData` were not IPC encodable or
decodable, so they were initialized with arbitrary values with UI-side compositing,
causing unexpected behavior and crashing.

This PR makes both of these encodable and decodable, provides equality comparison
operators for both, and provides all fields of both data types with sensible
initial values.

* Source/WebCore/page/scrolling/ScrollingCoordinatorTypes.h:
(WebCore::RequestedKeyboardScrollData::operator== const):
* Source/WebCore/platform/KeyboardScroll.h:
(WebCore::KeyboardScroll::operator== const):
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.cpp:
(ArgumentCoder<KeyboardScroll>::encode):
(ArgumentCoder<KeyboardScroll>::decode):
(WebKit::dump):

Canonical link: https://commits.webkit.org/257734@main
